### DISCUSSION TITLE
Javascript - Verify what volumes are under a specific Storage group

### DISCUSSION THREAD

**Discussion by johndanielk** [dell/PyU4V#196] [Q&A]
I'm still having (a little) trouble understanding everything related to the body in javascript

I need to know under Storage Group what volumes are under it
I think it's related to this URL, but not sure about it
https://developer.dell.com/apis/4458/versions/10.1.0/openapi_enhanced.json/paths/~1v1~1systems~1%7Bid%7D~1volumes/get

Thanks  guys

**Comment by rawstorage** [comment]
Hi John,
you can get this information from a number of places,   I'm presuming you are using PyU4V.  So i'll give a few examples.

import PyU4V

conn = PyU4V.U4VConn(
    username='smc',
    password='smc',
    server_ip='ipaddress',
    port='8443', array_id='000220200226',
    verify=False)

#Legacy Way
print(conn.provisioning.get_volumes_from_storage_group(
    storage_group_id='audit_test'))

#New Way
print(conn.storage_groups.get_storage_groups_details(filters=['id eq '
                                                              'audit_test'],
                                                     select=['volumes,volumes.resource_type,volumes.id,volumes.wwn,volumes.effective_wwn']))
#alternate new way to get more volume related information
print(conn.volumes.get_volumes_details(filters=['storage_groups.id eq audit_test']))

Sample output 
Legacy Way
['00270', '002C5']


New Way
{'storage_groups': [{'id': 'audit_test',
                     'volumes': [{'effective_wwn': '60000970000220200226533030323730',
                                  'id': '00270',
                                  'resource_type': 'volumes',
                                  'wwn': '60000970000220200226533030323730'},
                                 {'effective_wwn': '60000970000220200226533030324335',
                                  'id': '002C5',
                                  'resource_type': 'volumes',
                                  'wwn': '60000970000220200226533030324335'}]}]}


Alternate Way
{'volumes': [{'cap_cyl': 546.0,
              'cap_gb': 1.0,
              'cap_mb': 1024.0,
              'cap_tb': 0.0,
              'clones': [],
              'data_reduction_enabled': False,
              'data_reduction_ratio_to_one': 0.0,
              'effective_used_capacity_gb': 0.0,
              'effective_wwn': '60000970000220200226533030324335',
              'emulation': 'FBA',
              'encapsulated': False,
              'has_effective_wwn': False,
              'id': '002C5',
              'masking_views': [],
              'mobility_id_enabled': False,
              'nguid': '02265330303243350000976000022020',
              'num_of_clones': 0,
              'num_of_masking_views': 0,
              'num_of_rdf_infos': 1,
              'num_of_snapshots': 24,
              'num_of_storage_groups': 1,
              'rdf_infos': [{'local_rdf_group': {'id': '3'},
                             'remote_rdf_group': {'id': '3'},
                             'remote_volume': {'id': '0011F'}}],
              'resource_type': 'volumes',
              'snapshots': [{'id': '75840307201'},
                            {'id': '75838463233'},
                            {'id': '75836619265'},
                            {'id': '75834775041'},
                            {'id': '75832931073'},
                            {'id': '75831091969'},
                            {'id': '75829248001'},
                            {'id': '75827403777'},
                            {'id': '75825559809'},
                            {'id': '75823715841'},
                            {'id': '75821876737'},
                            {'id': '75820032769'},
                            {'id': '75818188545'},
                            {'id': '75816344577'},
                            {'id': '75814500609'},
                            {'id': '75812656385'},
                            {'id': '75810817537'},
                            {'id': '75808973313'},
                            {'id': '75807129345'},
                            {'id': '75805285377'},
                            {'id': '75803441153'},
                            {'id': '75801597185'},
                            {'id': '75799758337'},
                            {'id': '75797914113'}],
              'snapvx_source': True,
              'snapvx_target': False,
              'srp': {'id': 'SRP_1'},
              'ssid': '0xFFFFFFFF',
              'status': 'Ready',
              'storage_groups': [{'id': 'audit_test'}],
              'system': {'id': '000220200226'},
              'type': 'RDF1+TDEV',
              'unreducible_data_gb': 0.0,
              'wwn': '60000970000220200226533030324335'},
             {'cap_cyl': 546.0,
              'cap_gb': 1.0,
              'cap_mb': 1024.0,
              'cap_tb': 0.0,
              'clones': [],
              'data_reduction_enabled': False,
              'data_reduction_ratio_to_one': 0.0,
              'effective_used_capacity_gb': 0.0,
              'effective_wwn': '60000970000220200226533030323730',
              'emulation': 'FBA',
              'encapsulated': False,
              'has_effective_wwn': False,
              'id': '00270',
              'masking_views': [],
              'mobility_id_enabled': False,
              'nguid': '02265330303237300000976000022020',
              'num_of_clones': 0,
              'num_of_masking_views': 0,
              'num_of_rdf_infos': 1,
              'num_of_snapshots': 24,
              'num_of_storage_groups': 1,
              'rdf_infos': [{'local_rdf_group': {'id': '3'},
                             'remote_rdf_group': {'id': '3'},
                             'remote_volume': {'id': '00122'}}],
              'resource_type': 'volumes',
              'snapshots': [{'id': '75840307201'},
                            {'id': '75838463233'},
                            {'id': '75836619265'},
                            {'id': '75834775041'},
                            {'id': '75832931073'},
                            {'id': '75831091969'},
                            {'id': '75829248001'},
                            {'id': '75827403777'},
                            {'id': '75825559809'},
                            {'id': '75823715841'},
                            {'id': '75821876737'},
                            {'id': '75820032769'},
                            {'id': '75818188545'},
                            {'id': '75816344577'},
                            {'id': '75814500609'},
                            {'id': '75812656385'},
                            {'id': '75810817537'},
                            {'id': '75808973313'},
                            {'id': '75807129345'},
                            {'id': '75805285377'},
                            {'id': '75803441153'},
                            {'id': '75801597185'},
                            {'id': '75799758337'},
                            {'id': '75797914113'}],
              'snapvx_source': True,
              'snapvx_target': False,
              'srp': {'id': 'SRP_1'},
              'ssid': '0xFFFFFFFF',
              'status': 'Ready',
              'storage_groups': [{'id': 'audit_test'}],
              'system': {'id': '000220200226'},
              'type': 'RDF1+TDEV',
              'unreducible_data_gb': 0.0,
              'wwn': '60000970000220200226533030323730'}]}

If you are using the Native API call then you can use the REST Client downloadable from here https://github.com/rawstorage/reSTclient10/ 

A recent blog on using the new rest client

https://rawstorage.wordpress.com/2024/08/30/undocumented-feature-of-the-unisphere-for-powermax-rest-client-when-working-with-enhanced-api-calls-and-how-to-work-around/

**Comment by rawstorage** [comment]
updated comment with more readable JSON from pprint.

**Comment by rawstorage** [comment]
@johndanielk I've converted this issue to a discussion, let me know if you need anything else or if we can close.

**Comment by johndanielk** [comment]
Yes, I'm on it, please don't close the discussion
Of course I will update
Thank you very much

**Reply by rawstorage** [reply]
no worries.  I'm in europe so answers can be delayed, working US hours today though :)

**Comment by johndanielk** [comment]
Ok, I understood the native 
This is the script (For those who need it)

```
function getVolumeIds(restHost, symmetrixId, storageGroupId) {
   var url = "/univmax/restapi/101/sloprovisioning/symmetrix/" + symmetrixId + "/volume";  
    if (storageGroupId) {
        url += "?storageGroupId=" + storageGroupId;
    }

    System.log("Volume IDs URL: " + url);


    var req = restHost.createRequest("GET", url, null);

    var authHeader = "Basic " + basicAuto; //your Base64-encoded credentials
    req.setHeader("Authorization", authHeader);
    req.setHeader("Accept", "application/json");
    req.setHeader("Content-Type", "application/json");

    var res = req.execute();

    System.log("Response Status Code: " + res.statusCode);
    System.log("Response AllHeaders: " + res.getAllHeaders());

    if (res.statusCode != 200) {
        System.log("~~~~ Error: Failed to retrieve volumes ~~~");
        throw("Couldn't retrieve volumes: " + res.contentAsString);
    }

    var responseJson = JSON.parse(res.contentAsString);

    System.log("Volume IDs Response: " + JSON.stringify(responseJson));

    return responseJson.resultList.result;
}


var symmetrixId = 1234567890; // Replace with your actual symmetrixId
var storageGroupId = "My_Storage_Group"; // Optional, if you want to filter by storage group

var volumeIds = getVolumeIds(restHost, symmetrixId, storageGroupId);
System.log("Retrieved Volume IDs: " + JSON.stringify(volumeIds));
```

Now, my problem is to understand that I am not using Enhanced API ( I  have Pmax 8500 version 10.1)
I am sorry, but I did not understand this link - I cannot find the 
https://rawstorage.wordpress.com/2024/08/30/undocumented-feature-of-the-unisphere-for-powermax-rest-client-when-working-with-enhanced-api-calls-and-how-to-work-around/


this is my pic from my storage
![setting](https://github.com/user-attachments/assets/cb4a1ec2-486c-42de-bc65-d130c6bba9ee)

**Comment by rawstorage** [comment]
@johndanielk the Tool mentioned in this blog lets you easily traverse the meta data in the API documentation and quickly build query like REST calls using select and filter options the tool is available for download here https://github.com/rawstorage/reSTclient10/

https://ipaddress:8443/univmax/rest/v1/systems/000220200226/storage-groups?select=volumes,volumes.resource_type,volumes.id,volumes.wwn,volumes.effective_wwn&filter=id eq audit_test 

A Query like above can be built to get the details on all volumes in storage group audit_test

**Comment by rawstorage** [comment]
@johndanielk are you ok to close out this discussion?

**Comment by johndanielk** [comment]
Hi
Yes you can close
Thx
